pkg/controller/jobsync.go

```diff
@@ -0,0 +1,274 @@
+	GetJobFactory(owner metav1.Object) (JobFactory, error)
+
+	// GetOwnerJobSyncConditionStatus gets the status of the specified
+	// condition for the specified owner.
+	GetOwnerJobSyncConditionStatus(owner metav1.Object, conditionType JobSyncConditionType) bool
+
+	// DeepCopyOwner returns a deep copy of the owner object.
+	DeepCopyOwner(owner metav1.Object) metav1.Object
+
+	// SetOwnerJobSyncCondition sets the specified condition for the specified
+	// owner.
+	// The updateConditionCheck functions are used to determine whether
+	// the condition should be updated if there are changes to the reason or
+	// the message of the condition.
+	SetOwnerJobSyncCondition(
+		owner metav1.Object,
+		conditionType JobSyncConditionType,
+		status kapi.ConditionStatus,
+		reason string,
+		message string,
+		updateConditionCheck ...UpdateConditionCheck,
+	)
+
+	// OnJobCompletion is called when the processing job for the owner
+	// completes successfully.
+	OnJobCompletion(owner metav1.Object)
+
+	// OnJobFailure is called when the processing job for the owner fails.
+	OnJobFailure(owner metav1.Object)
+
+	// UpdateOwnerStatus updates the status of the owner from the original
+	// copy to the owner copy.
+	UpdateOwnerStatus(original, owner metav1.Object) error
+
+	// ProcessDeletedOwner processes an owner that has been marked for
+	// deletion.
+	ProcessDeletedOwner(owner metav1.Object) error
+}
+
+type jobSync struct {
+	jobControl JobControl
+	strategy   JobSyncStrategy
+	logger     log.FieldLogger
+}
+
+// NewJobSync creates a new JobSync.
+func NewJobSync(jobControl JobControl, strategy JobSyncStrategy, logger log.FieldLogger) JobSync {
+	return &jobSync{
+		jobControl: jobControl,
+		strategy:   strategy,
+		logger:     logger,
+	}
+}
+
+func (s *jobSync) Sync(key string) error {
+	logger := log.FieldLogger(s.logger.WithField("key", key))
+	startTime := time.Now()
+	logger.Debugln("Started syncing")
+	defer logger.WithField("duration", time.Since(startTime)).Debugln("Finished syncing")
+
+	owner, err := s.strategy.GetOwner(key)
+	if errors.IsNotFound(err) {
+		logger.Debugln("owner has been deleted")
+		s.jobControl.ObserveOwnerDeletion(key)
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	logger = loggerForOwner(s.logger, owner)
+
+	// Are we dealing with an owner marked for deletion
+	if owner.GetDeletionTimestamp() != nil {
+		logger.Debugf("DeletionTimestamp set")
+		return s.strategy.ProcessDeletedOwner(owner)
+	}
+
+	needsProcessing := s.strategy.DoesOwnerNeedProcessing(owner)
+
+	jobFactory, err := s.strategy.GetJobFactory(owner)
+	if err != nil {
+		return err
+	}
+
+	job, isJobNew, err := s.jobControl.ControlJobs(key, owner, needsProcessing, jobFactory)
+	if err != nil {
+		return err
+	}
+
+	switch {
+	// Owner is update to date
+	case !needsProcessing:
+		return nil
+	// New job has not been created, so an old job must exist. Set the owner
+	// to not processing as the old job is deleted.
+	case job == nil:
```

Comment:
Agreed about returning a more meaningful result. I briefly considered doing that for this PR but held off on it. I'll create an issue to track this to ensure it stays in my mind.

> Would we get to this case if expectations are not met yet? If so, are we doing the right thing by setting the owner to not processing?

This case happens when (1) there were no pending expectations and (2) there were existing jobs to delete. A new job was not created because we are waiting for the old jobs to be deleted first. The owner is set to not processing because a process job (that may have been active) was deleted.

Reply:
> This case happens when (1) there were no pending expectations and (2) there were existing jobs to delete. A new job was not created because we are waiting for the old jobs to be deleted first. The owner is set to not processing because a process job (that may have been active) was deleted.

If we have not met expectations, s.jobControl.ControlJobs returns nil, false, nil ... when do we catch that?

Reply:
That would be caught by the `job == nil` case, setting the owner to no provisioning, since there is not an active job.

Reply:
hmm, doesn't seem right. In the cluster controller we simply return when expectations have not been met yet

Reply:
Make sense. I'll re-work this to (1) make it more clear and (2) make it more correct.